File: 0893-all-nodes-distance-k-in-binary-tree/0893-all-nodes-distance-k-in-binary-tree.cpp
class Solution {
public:
    unordered_map<TreeNode*, TreeNode*> parentMap;

    void dfs(TreeNode* node, TreeNode* parent) {
        if (!node) return;
        parentMap[node] = parent;
        dfs(node->left, node);
        dfs(node->right, node);
    }

    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
        dfs(root, nullptr);  // build parent links

        unordered_map<TreeNode*, bool> visited;
        queue<TreeNode*> q;
        q.push(target);
        visited[target] = true;

        while (!q.empty() && k--) {
            int levelSize = q.size();
            for (int i = 0; i < levelSize; ++i) {
                TreeNode* node = q.front();
                q.pop();

                if (node->left && !visited[node->left]) {
                    q.push(node->left);
                    visited[node->left] = true;
                }
                if (node->right && !visited[node->right]) {
                    q.push(node->right);
                    visited[node->right] = true;
                }
                if (parentMap[node] && !visited[parentMap[node]]) {
                    q.push(parentMap[node]);
                    visited[parentMap[node]] = true;
                }
            }
        }

        vector<int> result;
        while (!q.empty()) {
            result.push_back(q.front()->val);
            q.pop();
        }

        return result;
    }
};



























// /**
//  * Definition for a binary tree node.
//  * struct TreeNode {
//  *     int val;
//  *     TreeNode *left;
//  *     TreeNode *right;
//  *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
//  * };
//  */
// class Solution {
// public:

//    unordered_map<TreeNode*,TreeNode*>parent;
//      void dfs(TreeNode* root , TreeNode* parent) {
//             if(!root){return;}
//             parent[root] = parent;
//            dfs(root->left,root);
//            dfs(root->right,root); 
//      }
//     vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
//         vector<int>visited(501,0);
//         queue<TreeNode*>Q;
//         Q.push(target);
//          while(!Q.empty() && k ) {
//                 k--;
//                 int level = Q.size();
//              for(int i = 0 ; i < level ; i++) {    
//                   TreeNode* node = Q.front();
//                   Q.pop();
//                  if(!visited[node->val]) {
//                     Q.push(parent[node]);
//                  }
//                   if(node->left){
//                      Q.push(node->left);
//                   }
//                   if(node->right){
//                     Q.push(node->right);
//                   } 
//              } 
//          }
//         vector<int>ans;
//         while(!Q.empty()){
//         TreeNode * node = Q.front();
//            Q.ppop();
//           ans.push_back(node->val);
//         }
//        return ans; 
//     }
// };

File: 0893-all-nodes-distance-k-in-binary-tree/0893-all-nodes-distance-k-in-binary-tree.java
class Solution {
    public void buildchildParentMap(TreeNode root, HashMap<TreeNode, TreeNode> childParent) {
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        while (!q.isEmpty()) {
            TreeNode node = q.poll();
            if (node.left != null) {
                q.add(node.left);
                childParent.put(node.left, node);
            }
            if (node.right != null) {
                q.add(node.right);
                childParent.put(node.right, node);
            }
        }
    }

    public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {
        HashMap<TreeNode, TreeNode> childParent = new HashMap<>();
        buildchildParentMap(root, childParent);
        HashMap<TreeNode, Boolean> Visited = new HashMap<>();
        Queue<TreeNode> q = new LinkedList<>();
        q.add(target);
        Visited.put(target, true);
        int level = 0;

        while (!q.isEmpty()) {
            int size = q.size();
            if (level == k) break;
            level++;
            for (int i = 0; i < size; i++) {
                TreeNode polled = q.poll();
                if (polled.left != null && !Visited.containsKey(polled.left)) {
                    Visited.put(polled.left, true);
                    q.add(polled.left);
                }
                if (polled.right != null && !Visited.containsKey(polled.right)) {
                    Visited.put(polled.right, true);
                    q.add(polled.right);
                }
                if (childParent.containsKey(polled) && !Visited.containsKey(childParent.get(polled))) {
                    TreeNode parent = childParent.get(polled);
                    Visited.put(parent, true);
                    q.add(parent);
                }
            }
        }

        List<Integer> ans = new ArrayList<>();
        while (!q.isEmpty()) ans.add(q.poll().val);
        return ans;
    }
}
